frontend/src/components/common/form/FastInput.tsx

```diff
@@ -0,0 +1,100 @@
+import React, { memo, useEffect } from 'react';
+import { Form, FormControlProps } from 'react-bootstrap';
+import { getIn, FormikProps } from 'formik';
+import { DisplayError } from './DisplayError';
+import { formikFieldMemo } from 'utils';
+
+type RequiredAttributes = {
+  /** The field name */
+  field: string;
+  /** formik state used for context and memo calculations */
+  formikProps: FormikProps<any>;
+};
+
+type OptionalAttributes = {
+  /** The form component label */
+  label?: string;
+  /** The underlying HTML element to use when rendering the FormControl */
+  as?: React.ElementType;
+  /** Short hint that describes the expected value of an <input> element */
+  placeholder?: string;
+  /** Adds a custom class to the input element of the <Input> component */
+  className?: string;
+  /** Whether the field is required. Shows an asterisk after the label. */
+  required?: boolean;
+  /** Specifies that the HTML element should be disabled */
+  disabled?: boolean;
+  /** Use React-Bootstrap's custom form elements to replace the browser defaults */
+  custom?: boolean;
+  /** Class name of the input wrapper */
+  outerClassName?: string;
+};
+
+// only "field" is required for <Input>, the rest are optional
+export type FastInputProps = FormControlProps & OptionalAttributes & RequiredAttributes;
+
+/**
+ * Formik-connected <Input> form control. Uses memo and cleanup inspired by
+ * https://jaredpalmer.com/formik/docs/api/fastfield
+ */
+export const FastInput: React.FC<FastInputProps> = memo(
+  ({
+    field,
+    label,
+    as: is, // `as` is reserved in typescript
+    placeholder,
+    className,
+    outerClassName,
+    required,
+    disabled,
+    custom,
+    formikProps: {
+      values,
+      errors,
+      touched,
+      handleChange,
+      handleBlur,
+      registerField,
+      unregisterField,
+    },
+    ...rest
+  }) => {
+    const error = getIn(errors, field);
+    const touch = getIn(touched, field);
+    const value = getIn(values, field);
+    const asElement: any = is || 'input';
+    useEffect(() => {
+      registerField(field, { validate: undefined });
+      return () => {
+        unregisterField(field);
+      };
+    }, []);
+
+    return (
+      <Form.Group
+        controlId={`input-${field}`}
+        className={[!!required ? 'required' : '', outerClassName ?? ''].join(' ')}
```

Comment:
The above logic has the unintended consequence to add a leading `space` to the **className** property when required is false -> `required blah` vs `[space] blah`. That's due to the use of `join(' ')`. 

We now see that effect in the changed snapshots as Jeremy pointed out.

Given that adding/removing CSS classes is a common concern of components as we build out frontend, would you consider using a library like [classnames](https://www.npmjs.com/package/classnames) ? It's is quite popular with over 5M downloads/week and there are plenty of resources/guides out there on how to use.

https://www.pluralsight.com/guides/applying-classes-conditionally-react

I've seen this in few places but will not add the same comment over and over. However, please fix or comment here on a solution suitable for all instances.

I'm okay if we push this change to a future PR as tech-debt

Reply:
sounds good to me.